File: server/gemini_service.py
import os
import google.generativeai as genai
from typing import Optional
import time
from models import ChatResponse, MessageResponse
from constants import SYSTEM_PROMPT
from db_service import DatabaseService

class ChatService:
    def __init__(self, db_service: DatabaseService):
        self.db_service = db_service
        api_key = "AIzaSyBfVeUyOi3Py6JL5B3pDGZKimGyK6c6kMg" or os.getenv("GEMINI_API_KEY") or os.getenv("API_KEY")
        if not api_key:
            raise ValueError("GEMINI_API_KEY environment variable is not set")
        
        genai.configure(api_key=api_key)
    
    async def send_message(
        self, 
        text: str, 
        session_id: Optional[str] = None
    ) -> ChatResponse:
        """
        Send a message and get AI response.
        Creates a new conversation if sessionId is not provided.
        """
        trimmed_message = text.strip()
        
        if not trimmed_message:
            raise ValueError("Message cannot be empty.")
        
        if len(trimmed_message) > 2000:
            raise ValueError("Message is too long. Please limit to 2000 characters.")
        
        # Generate session ID if not provided
        actual_session_id = session_id
        if not actual_session_id:
            current_time = int(time.time() * 1000)
            actual_session_id = f"sess_{current_time}"
            self.db_service.create_conversation(
                actual_session_id, 
                trimmed_message[:30] + "..."
            )
        
        # Save user message
        current_time = int(time.time() * 1000)
        user_msg = MessageResponse(
            id=f"msg_u_{current_time}",
            conversationId=actual_session_id,
            sender="user",
            text=trimmed_message,
            timestamp=current_time
        )
        self.db_service.save_message(user_msg)
        
        try:
            # Get conversation history
            history = self.db_service.get_messages_by_session(actual_session_id)
            
            # Prepare chat history for Gemini API (exclude the last message which is the current user message)
            chat_history = []
            for msg in history[:-1]:
                if msg.sender == "user":
                    chat_history.append({"role": "user", "parts": [msg.text]})
                else:
                    chat_history.append({"role": "model", "parts": [msg.text]})
            
            # Create model with system instruction
            # Use gemini-1.5-flash as default (works with free tier)
            # For paid tier, you can change to 'gemini-2.0-flash-exp' or 'gemini-1.5-pro'
            model = genai.GenerativeModel(
                'gemini-1.5-flash',
                system_instruction=SYSTEM_PROMPT
            )
            
            # Start chat session with history if there's any
            if chat_history:
                chat = model.start_chat(history=chat_history)
                response = chat.send_message(
                    trimmed_message,
                    generation_config={
                        "temperature": 0.7,
                        "max_output_tokens": 1000,
                    }
                )
            else:
                # First message in conversation
                response = model.generate_content(
                    trimmed_message,
                    generation_config={
                        "temperature": 0.7,
                        "max_output_tokens": 1000,
                    }
                )

            reply = response.text if response.text else "I'm sorry, I couldn't generate a response. Please try again."
            
            # Save AI response
            ai_time = int(time.time() * 1000)
            ai_msg = MessageResponse(
                id=f"msg_a_{ai_time}",
                conversationId=actual_session_id,
                sender="ai",
                text=reply,
                timestamp=ai_time
            )
            self.db_service.save_message(ai_msg)
            return ChatResponse(
                reply=reply,
                sessionId=actual_session_id
            )
            
        except Exception as error:
            error_msg = str(error)
            print(f"Gemini API Error: {error_msg}")  # Debug log
            
            if "API_KEY" in error_msg or "api_key" in error_msg or "authentication" in error_msg.lower():
                raise ValueError("System Error: Invalid configuration. Please check the API key.")
            
            # Handle quota/rate limit errors
            if "429" in error_msg or "quota" in error_msg.lower() or "rate limit" in error_msg.lower():
                raise ValueError("API quota exceeded. The free tier has limited requests. Please wait a moment and try again, or check your API plan at https://ai.google.dev/pricing")
            
            # Include the actual error message for debugging
            raise ValueError(f"The AI agent is currently unavailable: {error_msg}. Please try again in a moment.")

File: server-node/src/geminiService.ts
import { GoogleGenerativeAI } from '@google/generative-ai';
import { SYSTEM_PROMPT } from './constants';
import { ChatResponse, Message } from './types';
import { DatabaseService } from './dbService';

export class ChatService {
  private genAI: GoogleGenerativeAI;
  private dbService: DatabaseService;

  constructor(dbService: DatabaseService) {
    this.dbService = dbService;
    
    const apiKey = process.env.GEMINI_API_KEY || process.env.API_KEY;
    if (!apiKey) {
      throw new Error('GEMINI_API_KEY environment variable is not set');
    }

    this.genAI = new GoogleGenerativeAI(apiKey);
  }

  async sendMessage(text: string, sessionId?: string): Promise<ChatResponse> {
    const trimmedMessage = text.trim();

    if (!trimmedMessage) {
      throw new Error('Message cannot be empty.');
    }

    if (trimmedMessage.length > 2000) {
      throw new Error('Message is too long. Please limit to 2000 characters.');
    }

    // Generate session ID if not provided
    let actualSessionId = sessionId;
    if (!actualSessionId) {
      actualSessionId = `sess_${Date.now()}`;
      this.dbService.createConversation(
        actualSessionId,
        trimmedMessage.slice(0, 30) + '...'
      );
    }

    // Save user message
    const currentTime = Date.now();
    const userMsg: Message = {
      id: `msg_u_${currentTime}`,
      conversationId: actualSessionId,
      sender: 'user',
      text: trimmedMessage,
      timestamp: currentTime
    };
    this.dbService.saveMessage(userMsg);

    try {
      // Get conversation history
      const history = this.dbService.getMessagesBySession(actualSessionId);

      // Prepare chat history for Gemini API (exclude the last message which is the current user message)
      const chatHistory = history.slice(0, -1).map(msg => ({
        role: msg.sender === 'user' ? 'user' : 'model' as const,
        parts: [{ text: msg.text }]
      }));

      // Create model with system instruction
      // Use gemini-1.5-flash as default (works with free tier)
      // For paid tier, you can change to 'gemini-2.0-flash-exp' or 'gemini-1.5-pro'
      const model = this.genAI.getGenerativeModel({
        model: 'gemini-1.0-pro',
        systemInstruction: SYSTEM_PROMPT
      });

      let response;
      
      // Start chat session with history if there's any
      if (chatHistory.length > 0) {
        const chat = model.startChat({
          history: chatHistory
        });
        response = await chat.sendMessage(trimmedMessage, {
          generationConfig: {
            temperature: 0.7,
            maxOutputTokens: 1000
          }
        });
      } else {
        // First message in conversation
        response = await model.generateContent(trimmedMessage, {
          generationConfig: {
            temperature: 0.7,
            maxOutputTokens: 1000
          }
        });
      }

      const reply = response.response.text() || "I'm sorry, I couldn't generate a response. Please try again.";

      // Save AI response
      const aiTime = Date.now();
      const aiMsg: Message = {
        id: `msg_a_${aiTime}`,
        conversationId: actualSessionId,
        sender: 'ai',
        text: reply,
        timestamp: aiTime
      };
      this.dbService.saveMessage(aiMsg);

      return {
        reply,
        sessionId: actualSessionId
      };
    } catch (error: any) {
      const errorMsg = error?.message || String(error);
      console.error('Gemini API Error:', errorMsg);
      
      if (errorMsg.includes('API_KEY') || errorMsg.includes('api_key') || errorMsg.toLowerCase().includes('authentication')) {
        throw new Error('System Error: Invalid configuration. Please check the API key.');
      }
      
      // Handle quota/rate limit errors
      if (errorMsg.includes('429') || errorMsg.toLowerCase().includes('quota') || errorMsg.toLowerCase().includes('rate limit')) {
        throw new Error('API quota exceeded. The free tier has limited requests. Please wait a moment and try again, or check your API plan at https://ai.google.dev/pricing');
      }
      
      throw new Error(`The AI agent is currently unavailable: ${errorMsg}. Please try again in a moment.`);
    }
  }
}

